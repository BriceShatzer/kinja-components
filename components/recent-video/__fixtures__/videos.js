export default [
	{
		'id': '105201',
		'type': 'kinjavideo',
		'videoPage': true,
		'streamingUrl': 'https://kinja-vh.akamaihd.net/i/prod/105201/105201_,180p,270p,406p,720p,1080p,.mp4.csmil/master.m3u8',
		'poster': {
			'id': 'qa/3691511'
		},
		'postInfo': {
			'id': 1835238447,
			'thumbnailId': 'qa/3691511',
			'permalink': 'https://io9.gizmodo.com/the-first-order-watched-over-our-interview-in-star-wars-1835238447',
			'isVideo': true,
			'headline': 'The First Order Watched Over Our Interview in <i>Star Wars: Galaxy&#39;s Edge</i>',
			'publishTimeMillis': 1559676600447
		}
	},
	{
		'id': '101688',
		'type': 'kinjavideo',
		'videoPage': true,
		'streamingUrl': 'https://kinja-vh.akamaihd.net/i/prod/101688/101688_,180p,270p,406p,720p,1080p,.mp4.csmil/master.m3u8',
		'postInfo': {
			'id': 1835091510,
			'thumbnailId': 'qa/3690005',
			'permalink': 'https://io9.gizmodo.com/arya-stark-understood-death-but-chose-life-1835091510',
			'isVideo': true,
			'headline': 'Arya Stark Understood Death, But Chose Life',
			'publishTimeMillis': 1559318400510
		}
	},
	{
		'id': '114183',
		'type': 'kinjavideo',
		'videoPage': true,
		'streamingUrl': 'https://kinja-vh.akamaihd.net/i/prod/114183/114183_,180p,270p,406p,720p,1080p,.mp4.csmil/master.m3u8',
		'postInfo': {
			'id': 1835118531,
			'thumbnailId': 'xkb24d6jdoqbzy0f2b3h',
			'permalink': 'https://gizmodo.com/these-are-the-very-best-truly-wireless-earbuds-1835118531',
			'isVideo': true,
			'headline': 'We Found the Very Best Truly Wireless Earbuds',
			'publishTimeMillis': 1559311200531
		}
	},
	{
		'id': '105096',
		'type': 'kinjavideo',
		'videoPage': true,
		'streamingUrl': 'https://kinja-vh.akamaihd.net/i/prod/105096/105096_,180p,270p,406p,720p,1080p,.mp4.csmil/master.m3u8',
		'postInfo': {
			'id': 1835090598,
			'thumbnailId': 'qa/3688840',
			'permalink': 'https://io9.gizmodo.com/cersei-lannister-rose-to-the-top-of-a-crumbling-kingdom-1835090598',
			'isVideo': true,
			'headline': 'Cersei Lannister Rose to the Top of a Crumbling Kingdom',
			'publishTimeMillis': 1559150100598
		}
	},
	{
		'id': '109322',
		'type': 'kinjavideo',
		'videoPage': true,
		'streamingUrl': 'https://kinja-vh.akamaihd.net/i/prod/109322/109322_,180p,270p,406p,720p,1080p,.mp4.csmil/master.m3u8',
		'postInfo': {
			'id': 1835123686,
			'thumbnailId': 'qa/3689149',
			'permalink': 'https://gizmodo.com/googles-quest-to-build-the-perfect-one-tap-smartphone-c-1835123686',
			'isVideo': true,
			'headline': 'Google&#39;s Quest to Build the Perfect One-Tap Smartphone Camera',
			'publishTimeMillis': 1559744700686
		}
	}
];