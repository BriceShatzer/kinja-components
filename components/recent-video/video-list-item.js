/* @flow */

import React from 'react';
import styled, { css } from 'styled-components';
import { CSSTransition } from 'react-transition-group';
import Analytics, { type AnalyticsInjectedProps } from '../hoc/analytics';
import DateTime from '../../utils/DateTime';
import media from '../../style-utils/media';
import { EnsureDefaultTheme } from '../theme';
import VideoThumbnail from '../video-thumbnail';
import getEventLabel from './getEventLabel';

import type { Video } from 'module/video/videoRequest';

const Title = styled.div`
	font-weight: bold;
	font-size: 16px;
	line-height: 21px;

	${props => props.nowPlaying && css`
		color: ${props => props.theme.color.darkgray};
	`}

	${media.mediumUp`
		font-size: 19px;
		line-height: 24px;
		margin-top: 12px;

		${props => props.standalone && css`
			font-size: 24px;
			line-height: 31px;
			margin-top: 0;
			padding: 11px 0 3px;
		`}

	`}
`;

const Item = styled.a`
	position: relative;
	overflow: hidden;
	box-sizing: border-box;
	cursor: pointer;
	&& {
		color: ${props => props.theme.color.black};
	}

	&&:hover,
	&&:focus {
		color: ${props => props.theme.color.black};
		text-decoration: none;
	}

	&&:hover ${Title} {
		text-decoration: underline;
	}

	${props => props.type === 'carousel' && css`
		${media.smallOnly`
			margin-bottom: 15px;
			display: flex;
			align-items: flex-start;
		`}
	`}

	${props => props.type === 'endcard' && css`
		&& {
			color: ${props => props.theme.color.white};
		}

		&&:hover,
		&&:focus {
			color: ${props => props.theme.color.white};
		}

		&.transition-appear {
			opacity: 0;
		}

		&.transition-appear-active {
			opacity: 0.5;
			transition: opacity 300ms ease-in-out;
			transition-delay: ${props => props.index * 100}ms;
		}
	`}
`;

Item.displayName = 'Item';

const Metadata = styled.div`
	${media.smallOnly`
		width: 63%;
		flex: none;
		margin-left: 15px;
	`}
`;

const PublishDate = styled.div`
	color: ${props => props.theme.color.gray};
	font-size: 14px;
	line-height: 19px;
	margin-top: 3px;

	${props => props.type === 'endcard' && `
		opacity: 0.75;
	`}

	${props => props.standalone && css`
		margin-top: 0;
	`}
`;

const videoListType = {
	carousel: 'Carousel',
	endcard: 'End Card'
};

type Props = {
	clickHandler: (id: string) => void,
	video: Video,
	type: $Keys<typeof videoListType>,
	nowPlaying?: boolean,
	index: number,
	total: number,
	standalone?: boolean,
	clickOutToVideos?: boolean
} & AnalyticsInjectedProps;

export const VideoListItem = ({
	clickHandler,
	video,
	type,
	nowPlaying,
	index,
	total,
	standalone,
	clickOutToVideos,
	ga
}: Props) => {
	const permalink = clickOutToVideos ? video.postInfo && video.postInfo.permalink : null;

	const onClick = () => {
		if (!nowPlaying) {
			ga(
				'Video Player Interaction',
				`${videoListType[type]} click - position ${index + 1} of ${total}`,
				getEventLabel(video),
				{
					hitCallback: () => {
						if (clickOutToVideos) {
							window.location.href = permalink;
						}
					}
				}
			);

			if (!clickOutToVideos) {
				clickHandler(video.id);
			}
		}
	};

	return (
		<EnsureDefaultTheme>
			<CSSTransition
				classNames="transition"
				appear={type === 'endcard'}
				addEndListener={(node, done) => node.addEventListener('transitionend', done, false)}
			>
				<Item
					href={permalink}
					type={type}
					index={index}
					onClick={onClick}
				>
					<VideoThumbnail
						thumbnailId={video.postInfo.thumbnailId}
						nowPlaying={nowPlaying}
						standalone={standalone}
						width={type === 'endcard' ? 400 : 320}
						noLazy
					/>
					<Metadata>
						<Title
							nowPlaying={nowPlaying}
							standalone={standalone}
							dangerouslySetInnerHTML={{__html: video.postInfo.headline}}/>
						<PublishDate standalone={standalone} type={type}>
							{new DateTime({ timestamp: video.postInfo.publishTimeMillis, timezone: 'America/New_York'}).relativeDateTime}
						</PublishDate>
					</Metadata>
				</Item>
			</CSSTransition>
		</EnsureDefaultTheme>
	);
};

export default Analytics(VideoListItem);
