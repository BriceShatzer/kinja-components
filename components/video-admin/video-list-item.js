// @flow

import * as React from 'react';
import styled from 'styled-components';

import { defaultTheme } from '../theme';
import VideoPostLink from 'kinja-magma/models/VideoPostLink';
import SimpleImage from 'kinja-magma/models/SimpleImage';
import BlogAvatar from '../blog-avatar';
import Button from '../button19';
import TooltipWrapper from '../tooltip/tooltip-wrapper';
import CheckmarkIcon from '../icon19/Checkmark';
import ClockIcon from '../icon19/Clock';
import IssueIcon from '../icon19/Issue';
import ErrorIcon from '../icon19/Error';
import PencilIcon from '../icon19/Pencil';
import ImageIcon from '../icon19/Image';
import CopyIcon from '../icon19/Copy';
import { LazyResponsiveImage, MAX_LEFT_OF_HEADLINE_IMAGE_WIDTH } from '../elements/image';
import placeholderImage from 'kinja-images/placeholderImage';

import { getTimestamp, getDuration } from './utils';
import { EditModal, CustomPosterModal } from '../video-uploader';
import { ListRow, ListColumn } from './video-list';
import CircularProgressBar from './circular-progress-bar';

import type { Program, VideoMetadataFields } from 'kinja-magma/api/video';
import type { VideoMetaJSON } from 'kinja-magma/models/VideoMeta';
import type { Locale } from 'kinja-magma/models/Locale';
import type { Image } from '../types';

type Props = {
	videoId: string,
	network: string,
	title: string,
	publishedTime?: string,
	duration?: number,
	posts?: Array<VideoPostLink>,
	poster?: SimpleImage,
	locale?: Locale,
	enableCaptionUpload: boolean,
	progress: number,
	isUploaded: boolean,
	isMetaSaved: boolean,
	isTranscoded: boolean,
	errorMessage?: string,
	uploadImage: (string | File) => Promise<*>,
	getPrograms: () => Promise<Array<Program>>,
	getVideoInfo: (videoId: string, bustCache: boolean) => Promise<VideoMetaJSON>,
	onVideoUpdateSubmit: (videoId: string, fields: VideoMetadataFields, captionFile: ?File) => Promise<*>,
	onCustomPosterChange: (videoId: string, image: Image) => Promise<*>,
	onCopyURLClick: (videoId: string) => void
};

export const CutoffText = styled.span`
	display: block;
	overflow: hidden;
	text-overflow: ellipsis;
	max-width: 100%;

	> a {
		color: unset;
		transition: color 250ms ease;

		&:hover {
			color: ${defaultTheme.color.gray};
			text-decoration: none;
		}
	}
`;

export const EmphasizedText = styled(CutoffText)`
	color: ${defaultTheme.color.darkgray};
	font-style: italic;
	font-weight: ${props => props.strong ? '400' : '300'};
`;

export const PostList = styled.ul`
	font-size: inherit;
	margin-left: 0;
	margin-bottom: 0;
	overflow: hidden;
`;

export const PostItem = styled.li`
	display: flex;
	align-items: center;
	line-height: 1;
	> span:first-child {
		flex: none;
		margin-right: 0.5rem;
	}

	&:not(:first-child) {
		margin-top: 5px;
	}
`;

export const ActionButton = styled(Button)`
	margin-right: 4px;
	&:last-child {
		margin-right: 0;
	}
`;

export const getLogoName = (network?: string): ?string => {
	switch (network) {
		case 'a.v. club':
			return 'avclub';
		case 'the root':
			return 'theroot';
		case 'the onion':
			return 'theonion';
		case 'the takeout':
			return 'thetakeout';
		case 'deadspin':
		case 'clickhole':
		case 'earther':
		case 'gizmodo':
		case 'jalopnik':
		case 'jezebel':
		case 'kotaku':
		case 'lifehacker':
		case 'splinter':
			return network;
		default:
			return null;
	}
};

export const renderStatusIcon = ({
	isUploaded,
	isMetaSaved,
	isTranscoded,
	errorMessage,
	progress
}: {
	isUploaded: boolean,
	isMetaSaved: boolean,
	isTranscoded: boolean,
	errorMessage?: ?string,
	progress: number
}): React.Element<typeof TooltipWrapper> => {
	let icon = null;
	let message;
	if (errorMessage) {
		message = errorMessage;
		icon = <ErrorIcon description={errorMessage} />;
	} else if (isUploaded) {
		if (isTranscoded) {
			message = 'Your video is ready.';
			icon = <CheckmarkIcon />;
		} else if (isMetaSaved) {
			message = 'This video is being transcoded.\nYou can embed it into posts now.';
			icon = <ClockIcon />;
		} else {
			message = 'Video uploaded but missing required info.\nEdit now to complete upload.';
			icon = <IssueIcon />;
		}
	} else if (isMetaSaved) {
		message = `This video is being uploaded, ${progress}% complete.`;
		icon = <CircularProgressBar progress={progress} />;
	} else {
		message = 'This video is being uploaded.\nAdd missing required info.';
		icon = <CircularProgressBar progress={progress} />;
	}

	return (
		<TooltipWrapper>
			{({toggleTooltip}) => {
				return (
					<div
						data-shouldshowtooltip={true}
						data-tooltipname={message}
						data-tooltipoffset={25}
						onMouseEnter={toggleTooltip}
						onMouseLeave={toggleTooltip}>
						{icon}
					</div>
				);
			}}
		</TooltipWrapper>
	);
};

export const renderTitle = (title?: string) =>
	title ? title : <EmphasizedText strong>This video has no title yet</EmphasizedText>;

export const renderPosts = (posts: Array<VideoPostLink> = [], blogName?: ?string) => {
	if (posts && posts.length > 0) {
		return (
			<PostList>
				{posts.map(post =>
					<PostItem key={post.id}>
						{blogName && <BlogAvatar name={blogName} size={21} />}
						<CutoffText>
							<a href={post.permalink} target="_blank" rel="noopener noreferrer" dangerouslySetInnerHTML={{ __html: post.headline }} />
						</CutoffText>
					</PostItem>
				)}
			</PostList>
		);
	}

	return <EmphasizedText>This video doesn&#39;t appear in any post</EmphasizedText>;
};

export default function VideoListItem(props: Props) {
	const {
		videoId,
		network,
		title,
		publishedTime,
		posts,
		poster,
		duration,
		locale = 'en-US',
		isUploaded,
		isMetaSaved,
		isTranscoded,
		errorMessage,
		progress,
		enableCaptionUpload,
		uploadImage,
		getPrograms,
		getVideoInfo,
		onVideoUpdateSubmit,
		onCustomPosterChange,
		onCopyURLClick
	} = props;
	const [isEditing, setIsEditing] = React.useState(false);
	const [isUploadingPoster, setIsUploadingPoster] = React.useState(false);
	const blogName = getLogoName(network);

	const onCustomPosterUpload = (image: Image): void => {
		onCustomPosterChange(videoId, image)
			.then(() => setIsUploadingPoster(false));
	};

	const renderEditModal = () => (
		<EditModal
			onCancel={() => setIsEditing(false)}
			onSubmit={(id, fields, caption) => onVideoUpdateSubmit(id, fields, caption).then(() => setIsEditing(false))}
			getPrograms={getPrograms}
			onProgramError={() => {}}
			getVideoInfo={() => getVideoInfo(videoId, true)}
			enableCaptionUpload={enableCaptionUpload}
		/>
	);

	const renderCustomPosterModal = () => (
		<CustomPosterModal
			imageUploader={uploadImage}
			onChange={onCustomPosterUpload}
			onCancel={() => setIsUploadingPoster(false)}
		/>
	);

	return (
		<ListRow>
			<ListColumn>
				{poster ? <LazyResponsiveImage
					alt={title}
					ariaLabel={title}
					format={poster.format}
					id={poster.id}
					width={MAX_LEFT_OF_HEADLINE_IMAGE_WIDTH}
					transform="KinjaCenteredLargeAuto"
					croppedImage
					noLazy
					relative
				/> : <img src={placeholderImage(blogName)} />}
			</ListColumn>
			<ListColumn><CutoffText>{renderTitle(title)}</CutoffText></ListColumn>
			<ListColumn centered>
				{renderStatusIcon({
					isUploaded,
					isMetaSaved,
					isTranscoded,
					errorMessage,
					progress
				})}
			</ListColumn>
			<ListColumn>{publishedTime && getTimestamp(publishedTime, locale)}</ListColumn>
			<ListColumn centered>{duration && getDuration(duration)}</ListColumn>
			<ListColumn>
				{renderPosts(posts, blogName)}
			</ListColumn>
			<ListColumn>
				<ActionButton
					icon={<PencilIcon />}
					variant="tertiary"
					onClick={() => setIsEditing(true)}
				/>
				<ActionButton
					icon={<ImageIcon />}
					variant="tertiary"
					onClick={() => setIsUploadingPoster(true)}
					isSmall
				/>
				<ActionButton
					icon={<CopyIcon />}
					label="Copy URL"
					labelPosition="after"
					onClick={onCopyURLClick}
					isSmall
				/>
			</ListColumn>
			{isEditing && renderEditModal()}
			{isUploadingPoster && renderCustomPosterModal()}
		</ListRow>
	);
}
