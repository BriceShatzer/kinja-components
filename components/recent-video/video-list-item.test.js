import * as React from 'react';
import { shallow } from 'enzyme';

import { VideoListItem } from './video-list-item';
import videos from './__fixtures__/videos';

const noop = () => { };

const mockVideo = videos[0];

const stubElement = ({
	video = mockVideo,
	onClick = noop,
	type = 'carousel',
	clickOutToVideos = false,
	index = 0,
	total = 1,
	ga = noop
} = {}) => (
	<VideoListItem
		video={video}
		clickHandler={onClick}
		type={type}
		clickOutToVideos={clickOutToVideos}
		index={index}
		total={total}
		ga={ga}
	/>
);

describe('<VideoListItem />', () => {
	it('should render by default', () => {
		const wrapper = shallow(stubElement());
		expect(wrapper).toMatchSnapshot();
	});

	it('should call the onVideoClick prop when clicked', () => {
		const onClick = jest.fn();
		const wrapper = shallow(stubElement({
			onClick
		}));

		wrapper.find('Item').first().simulate('click');
		expect(onClick).toHaveBeenCalledWith(mockVideo.id);
	});

	it('should not call the onVideoClick prop when clicked, if element is clickOutToVideos', () => {
		const onClick = jest.fn();
		const wrapper = shallow(stubElement({
			onClick,
			clickOutToVideos: true
		}));

		wrapper.find('Item').first().simulate('click');
		expect(onClick).not.toHaveBeenCalled();
	});

	it('should have link href if it\'s clickOutToVideos', () => {
		const onClick = jest.fn();
		const wrapper = shallow(stubElement({
			onClick,
			clickOutToVideos: true
		}));

		const href = wrapper.find('Item').first().prop('href');
		expect(href).toMatchSnapshot();
	});
});
