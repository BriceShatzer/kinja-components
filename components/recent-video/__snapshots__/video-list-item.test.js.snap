// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<VideoListItem /> should have link href if it's clickOutToVideos 1`] = `"https://io9.gizmodo.com/the-first-order-watched-over-our-interview-in-star-wars-1835238447"`;

exports[`<VideoListItem /> should render by default 1`] = `
<EnsureDefaultTheme>
  <CSSTransition
    addEndListener={[Function]}
    appear={false}
    classNames="transition"
  >
    <Item
      href={null}
      index={0}
      onClick={[Function]}
      type="carousel"
    >
      <VideoThumbnail
        noLazy={true}
        thumbnailId="qa/3691511"
        width={320}
      />
      <video-list-item__Metadata>
        <video-list-item__Title
          dangerouslySetInnerHTML={
            Object {
              "__html": "The First Order Watched Over Our Interview in <i>Star Wars: Galaxy&#39;s Edge</i>",
            }
          }
        />
        <video-list-item__PublishDate
          type="carousel"
        >
          6/04/19 3:30PM
        </video-list-item__PublishDate>
      </video-list-item__Metadata>
    </Item>
  </CSSTransition>
</EnsureDefaultTheme>
`;
